added content card inbox tutorial + small scrolly fix (#10190)
braze-inc/braze-docs: _docs/_developer_guide/content_cards/content_card_inbox.md

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        cards.count
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let card = cards[indexPath.row]
        let cell = tableView.dequeueReusableCell(withIdentifier: "CardCell", for: indexPath)

        // Work with the content card's title and description
        cell.textLabel?.numberOfLines = 2
        cell.textLabel?.text = [card.title, card.description].compactMap { $0 }.joined(separator: "\n")

        // Load image (if present)
        if let imageURL = card.imageURL {
            // custom logic
        }

        card.logImpression(using: AppDelegate.braze)
        return cell
    }

    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let card = cards[indexPath.row]
        card.logClick(using: AppDelegate.braze)
        if let url = card.clickAction?.url {
            UIApplication.shared.open(url, options: [:], completionHandler: nil)
        }
        tableView.deselectRow(at: indexPath, animated: true)
    }

> 

Reviewer: We actually allow logging [multiple times for content cards](https://github.com/braze-inc/swift-sdk/blob/0a846593fcd4a33130d8e47fd3b59c64a311e3ce/Tests/BrazeKitTests/ContentCards/ContentCardContextTests.swift#L34) so the proposed `tableView(_:willDisplay:forRowAt:)` snippet wouldn't actually work as intended (it would log each time the cell displays so it would have duplicates). I need to get back to you on how we'd recommend logging impressions with custom Content Cards UI, since our default UI has some pretty sophisticated logic

---
nav_title: "Tutorial: Content Card Inboxes"
article_title: "Tutorial: Making an Inbox with Content Cards"
description: ""
page_order: 1
layout: scrolly
---

# Tutorial: Making an Inbox with Content Cards

> Follow along with the sample code in this tutorial to build an inbox with Braze content cards.

{% sdktabs %}
{% sdktab android %}
{% multi_lang_include developer_guide/prerequisites/android.md %}

## Making an Inbox with Content Cards for Android

{% multi_lang_include developer_guide/_shared/tutorial_feedback.md %}

{% scrolly %}

```kotlin file=MainApplication.kt
import android.app.Application
import com.braze.Braze
import com.braze.support.BrazeLogger
import com.braze.configuration.BrazeConfig
import android.util.Log

class ContentCardsApplication : Application() {
    override fun onCreate() {
        super.onCreate()

        // Turn on verbose Braze logging

>         BrazeLogger.logLevel = Log.VERBOSE

Reviewer 2: ```suggestion
        BrazeLogger.enableVerboseLogging()
```

        // Configure Braze with your SDK key & endpoint
        val config = BrazeConfig.Builder()
            .setApiKey("1ff5603a-4c31-40c3-bb84-e2f56d414ae9")

>             .setCustomEndpoint("sondheim.braze.com")

Author: ```suggestion
            .setCustomEndpoint("YOUR_API_ENDPOINT")
```

}
```

```xml file=content_card_inbox.xml
<?xml version="1.0" encoding="utf-8"?>
<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
        android:orientation="vertical"
        android:layout_width="match_parent"
        android:layout_height="match_parent">

        <TextView
            android:id="@+id/inboxHeader"
            android:layout_width="match_parent"
            android:layout_height="wrap_content"
            android:text="Message Inbox"
            android:textStyle="bold"
            android:textSize="20sp"
            android:paddingStart="16dp"
            android:paddingEnd="16dp"
            android:paddingTop="12dp"
            android:paddingBottom="8dp" />

        <androidx.recyclerview.widget.RecyclerView
            android:id="@+id/contentCardsRecyclerView"
            android:layout_width="match_parent"
            android:layout_height="0dp"
            android:layout_weight="1" />
    </LinearLayout>

```

!!step
lines-MainApplication.kt=12

#### 1. Enable debugging (optional)

To make troubleshooting easier while developing, consider enabling debugging.

!!step
lines-content_card_inbox.xml=1-24

#### 2. Create the UI View

> We're using Android's `RecyclerView` to display Content Cards in this tutorial, but we recommend build a UI with calsses and components that suit your use case(s). Braze provides UI by default, but here we create a custom view to have full control over the appearance and behavior.

Reviewer 3: ```suggestion
We're using Android's `RecyclerView` to display Content Cards in this tutorial, but we recommend building a UI with classes and components that suit your use case(s). Braze provides UI by default, but here we create a custom view to have full control over the appearance and behavior.
```

> We're using Android's `RecyclerView` to display Content Cards in this tutorial, but we recommend build a UI with calsses and components that suit your use case(s). Braze provides UI by default, but here we create a custom view to have full control over the appearance and behavior.

Reviewer 4: Might be useful to link to the Android docs here: 

```suggestion
We're using Android's [`RecyclerView`](https://developer.android.com/develop/ui/views/layout/recyclerview) to display Content Cards in this tutorial, but we recommend build a UI with calsses and components that suit your use case(s). Braze provides UI by default, but here we create a custom view to have full control over the appearance and behavior.
```

!!step
lines-ContentCardInboxActivity.kt=29-35,40-42,44

#### 3. Subscribe to & Refresh Content Cards

Use [`subscribeToContentCardsUpdates`](<https://braze-inc.github.io/braze-android-sdk/kdoc/braze-android-sdk/com.braze/-i-braze/subscribe-to-content-cards-updates.html?query=abstract%20fun%20subscribeToContentCardsUpdates(subscriber:%20IEventSubscriber%3CContentCardsUpdatedEvent%3E)/>) to allow your UI to respond when new content cards are available. Here, subscribers are registered and removed within the activity lifecycle hooks.

!!step
lines-ContentCardInboxActivity.kt=73-84
#### 4. Build a Custom Inbox UI

> Using the content card [attributes](<https://braze-inc.github.io/braze-android-sdk/kdoc/braze-android-sdk/com.braze.models.cards/-card/index.html/>) such as `title`, `description`, and `url` allows you to build content cards to match your specific UI requirements. In this case, we're building an inbox with Android's native `RecyclerView`.

Reviewer 2: This link is dead. Maybe:

```
https://braze-inc.github.io/braze-android-sdk/kdoc/braze-android-sdk/com.braze.models.cards/-card/index.html?query=open%20class%20Card(jsonObject:%20JSONObject,%20brazeManager:%20IBrazeManager?,%20cardStorage:%20ICardStorageProvider%3C*%3E?,%20cardAnalytics:%20ICardAnalyticsProvider?)%20:%20IPutIntoJson%3CJSONObject?%3E
```

Reviewer 3: https://braze-inc.github.io/braze-android-sdk/kdoc/braze-android-sdk/com.braze.models.cards/-card/index.html works for me

!!step
lines-ContentCardInboxActivity.kt=90,93
#### 5. Track Impressions & Clicks

> You can log impressions and clicks using the [`logImpressions`](<https://braze-inc.github.io/braze-android-sdk/kdoc/braze-android-sdk/com.braze.models.cards/-card/log-impression.html/>) and [`logClick`](<https://braze-inc.github.io/braze-android-sdk/kdoc/braze-android-sdk/com.braze.models.cards/-card/log-click.html/>) methods available for content cards.

Reviewer 2: both links are dead

https://braze-inc.github.io/braze-android-sdk/kdoc/braze-android-sdk/com.braze.models.cards/-card/log-impression.html?query=fun%20logImpression():%20Boolean

https://braze-inc.github.io/braze-android-sdk/kdoc/braze-android-sdk/com.braze.models.cards/-card/log-click.html

Reviewer 3: both of these  in the original PR work for me

> Impressions should only be logged once when a card is viewed by the user. Here, we use a naive machanism to guard agaisnt duplicate logs with a per-card flag. Note that you may need to think through the view lifecycle of your app, as well as use case, so ensure impressions are logged correctly.

Reviewer 4: ```suggestion
Impressions should only be logged once when a card is viewed by the user. Here, we use a naive mechanism to guard against duplicate logs with a per-card flag. Note that you may need to think through the view lifecycle of your app, as well as use case, so ensure impressions are logged correctly.
```
